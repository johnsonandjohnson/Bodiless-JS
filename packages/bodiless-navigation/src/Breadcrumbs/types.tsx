/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable jsx-a11y/anchor-is-valid */
import { ComponentType, HTMLProps } from 'react';
import type { WithNodeProps } from '@bodiless/core';
import type { DesignableComponentsProps } from '@bodiless/fclasses';
import type { BreadcrumbItemType as BreadcrumbStoreItemType } from './BreadcrumbStore';

type BreadcrumbsComponents = {
  StartingTrail: ComponentType<any>,
  Separator: ComponentType<HTMLProps<HTMLSpanElement>>,
  Wrapper: ComponentType<HTMLProps<HTMLUListElement>>,
  Item: ComponentType<HTMLProps<HTMLLIElement> & {
    position: number;
    isCurrentPage: boolean;
  }>,
  Title: ComponentType<any>,
  FinalTrail: ComponentType<any>,
};

type CleanBreadcrumbItemType = {
  uuid: string | number;
  position: number;
  isCurrentPage: boolean;
} & WithNodeProps;

/**
 * contains breadcrumb item public properties
*/
type BreadcrumbItemType = Pick<BreadcrumbStoreItemType, 'uuid' | 'title' | 'link' | 'isFirst' | 'hasPath'>;

/**
 * reduces items retrieved from breadcrumb store
 *
 * @param items - a list of items retrieved from store
 * @param props - props passed to breadcrumb component
 *
 * @returns uuids - a collection of breadcrumb item uuids
 */
type BreadcrumbStoreItemsReducer = (
  items: BreadcrumbItemType[],
  props?: Pick<BreadcrumbsProps, 'hasStartingTrail' | 'hasFinalTrail'>,
) => string[];

type BreadcrumbsProps = DesignableComponentsProps<BreadcrumbsComponents> & {
  /**
   * whether starting custom item is enabled and should be rendered
   * default: disabled
   */
  hasStartingTrail?: boolean | (() => boolean),
  /**
   * list of breadcrumb items to render
   */
  items?: CleanBreadcrumbItemType[],
  /**
   * whether final custom item is enabled and should be rendered
   */
  hasFinalTrail?: boolean | (() => boolean),
  /**
   * allows to reduce items retrieved from breadcrumb store
   * default: firstItemHomeLinkReducer
   */
  itemsReducer?: BreadcrumbStoreItemsReducer,
  /**
   * whether the last breadcrumb item should not be rendered as a link
   * default: false
   */
  renderLastItemWithoutLink?: boolean | (() => boolean),
} & WithNodeProps;

type CleanBreadcrumbsProps = Omit<BreadcrumbsProps, 'itemsReducer'>;

export type {
  BreadcrumbsComponents,
  BreadcrumbsProps,
  CleanBreadcrumbsProps,
  BreadcrumbItemType,
  CleanBreadcrumbItemType,
  BreadcrumbStoreItemsReducer,
};
