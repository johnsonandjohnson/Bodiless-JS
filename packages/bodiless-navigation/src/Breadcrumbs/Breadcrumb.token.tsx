/**
 * Copyright Â© 2021 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable import/prefer-default-export */
import { stylable, withDesign, asToken } from '@bodiless/fclasses';
import type { TokenDef } from '@bodiless/fclasses';

/**
 * Makes all Breadcrumb design components stylable.
 */
export const asStylableBreadcrumbs = withDesign({
  StartingTrail: stylable,
  Separator: stylable,
  Wrapper: stylable,
  BreadcrumbItem: stylable,
  Title: stylable,
  FinalTrail: stylable,
  Item: stylable,
});

/**
 * Applies supplied tokenDefs to all breadcrumb items.
 *
 * @returns Token - a token that applies provided tokenDefs to all breadcrumb items.
 */
export const withBreadcrumbItemToken = (...tokenDefs: TokenDef<any>[]) => withDesign({
  StartingTrail: asToken(...tokenDefs),
  FinalTrail: asToken(...tokenDefs),
  Title: asToken(...tokenDefs),
});
