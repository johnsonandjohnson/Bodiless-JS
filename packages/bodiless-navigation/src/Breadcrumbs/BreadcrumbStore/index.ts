export * from './BreadcrumbStore bl-edit';
