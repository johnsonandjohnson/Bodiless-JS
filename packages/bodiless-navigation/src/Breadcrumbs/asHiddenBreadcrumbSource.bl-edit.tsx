import React from 'react';
import { renderToPipeableStream } from 'react-dom/server';
import { HOC } from '@bodiless/fclasses';
import { useNode, NodeProvider } from '@bodiless/data';
import { useBreadcrumbStore, BreadcrumbStoreProvider } from './BreadcrumbStoreProvider';
/**
 * @private
 *
 * Wrap a breadcrumb source which is rendered only on the server to populate
 * the store before any components which need breadcrumb store data are
 * rendered.
 *
 * @param Component
 */
const asHiddenBreadcrumbSource: HOC = Component => {
  const AsHiddenBreadcrumbSource = (props: any) => {
    const store = useBreadcrumbStore();
    const { node } = useNode();

    renderToPipeableStream(
      <NodeProvider node={node}>
        <BreadcrumbStoreProvider store={store}>
          <Component {...props} />
        </BreadcrumbStoreProvider>
      </NodeProvider>,
    );
    return null;
  };
  return AsHiddenBreadcrumbSource;
};

export default asHiddenBreadcrumbSource;
