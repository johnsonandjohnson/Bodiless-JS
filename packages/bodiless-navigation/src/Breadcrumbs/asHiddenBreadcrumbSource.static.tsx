import React from 'react';
import { HOC } from '@bodiless/fclasses';

const asHiddenBreadcrumbSource: HOC = Component => (props: any) => <Component {...props} />;

export default asHiddenBreadcrumbSource;
