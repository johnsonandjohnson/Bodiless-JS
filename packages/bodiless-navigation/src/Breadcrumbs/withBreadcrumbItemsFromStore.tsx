/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable jsx-a11y/anchor-is-valid */
import React, { ComponentType } from 'react';
import { useNode } from '@bodiless/core';
import { useBreadcrumbStore } from './BreadcrumbStoreProvider';

import type {
  BreadcrumbsProps,
  BreadcrumbItemType,
  CleanBreadcrumbItemType,
} from './types';

/**
 * removes first item from the trail
 * when there is a custom starting trail and
 * when the first store item has frontpage path
 *
 * @param items - breadcrumb store items
 * @param props - breadcrumb component props
 *
 * @returns uuids - a list of item uuids
 */
const firstItemHomeLinkReducer = (
  items: BreadcrumbItemType[],
  { hasStartingTrail }: Pick<BreadcrumbsProps, 'hasStartingTrail'>,
) => items
  .filter(item => !(hasStartingTrail && item.isFirst() && item.hasPath('/')))
  .map(item => item.uuid);

/**
 * HOC that populates a breadcrumb based component with data from breadcrumb store.
 * @param Component a breadcrumb based component
 */
const withBreadcrumbItemsFromStore = (Component: ComponentType<BreadcrumbsProps>) => {
  const WithBreadcrumbItemsFromStore = (props: BreadcrumbsProps) => {
    const {
      nodeCollection,
      hasStartingTrail = false,
      hasFinalTrail = false,
      itemsReducer = firstItemHomeLinkReducer,
      renderLastItemWithoutLink = true,
      ...rest
    } = props;
    const store = useBreadcrumbStore();
    if (store === undefined) return <Component {...props} />;
    const { node } = useNode(nodeCollection);
    const basePath = node.path;
    const items = itemsReducer(store.breadcrumbTrail, { hasStartingTrail })
      .map(uuid => store.getItem(uuid))
      // map items retrieved from store
      // into items expected by base breadcrumb component
      /* eslint-disable @typescript-eslint/indent */
      // eslint throws an indentation error for lines inside reduce body
      // automatic eslint fix brings code to unreadable state
      // probably that is an eslint plugin issue
      // the disabled rule is enabled back after reduce
      .reduce<CleanBreadcrumbItemType[]>(
        (prev, current, index) => {
          if (current === undefined) return prev;
          const titleNodePath = current.title.nodePath.replace(`${basePath}$`, '');
          // @todo probably a better way to get the nodeKey...
          const temp = titleNodePath.split('$');
          const nodeKey = temp.slice(0, temp.length - 1).join('$');
          prev.push({
            uuid: current.uuid,
            nodeKey,
            nodeCollection,
            position: index + 1,
            isCurrentPage: current.isLast() && store.hasCurrentPageItem(),
          });
          return prev;
        }, [],
      );
    /* eslint-enable @typescript-eslint/indent */
    const hasFinalTrail$0 = typeof hasFinalTrail === 'function' ? hasFinalTrail() : hasFinalTrail;
    const hasFinalTrail$1 = hasFinalTrail$0 && !store.hasCurrentPageItem();
    const lastItemWithoutLink = typeof renderLastItemWithoutLink === 'function'
      ? renderLastItemWithoutLink()
      : renderLastItemWithoutLink;
    const props$1 = {
      ...rest,
      items,
      hasFinalTrail: hasFinalTrail$1,
      hasStartingTrail,
      renderLastItemWithoutLink: lastItemWithoutLink
        && !hasFinalTrail$1
        && store.hasCurrentPageItem(),
    };
    return <Component {...props$1} />;
  };
  return WithBreadcrumbItemsFromStore;
};

export default withBreadcrumbItemsFromStore;
export {
  firstItemHomeLinkReducer,
};
