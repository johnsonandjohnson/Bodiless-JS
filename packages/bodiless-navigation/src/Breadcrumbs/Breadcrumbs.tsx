/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable jsx-a11y/anchor-is-valid */
import React, { ComponentType } from 'react';
import { withoutProps, withNode } from '@bodiless/core';
import type { WithNodeProps } from '@bodiless/core';
import {
  asComponent, designable, addProps, Fragment, withDesign, replaceWith,
} from '@bodiless/fclasses';
import { observer } from 'mobx-react-lite';
import { flowRight } from 'lodash';
import { asStylableBreadcrumbs } from './Breadcrumb.token';

import withBreadcrumbItemsFromStore from './withBreadcrumbItemsFromStore';
import withBreadcrumbsSD from './withBreadcrumbsStructuredData';
import MenuTitle from '../Menu/MenuTitles';

import type {
  BreadcrumbsComponents,
  CleanBreadcrumbsProps,
  CleanBreadcrumbItemType,
} from './types';

const ItemNodeProvider = withNode(Fragment) as ComponentType<WithNodeProps>;

const BreadcrumbsClean$ = (props: CleanBreadcrumbsProps) => {
  const {
    hasStartingTrail = false,
    components,
    items = [],
    hasFinalTrail = false,
    renderLastItemWithoutLink = false,
  } = props;
  const hasStartingTrail$ = typeof hasStartingTrail === 'function' ? hasStartingTrail() : hasStartingTrail;
  const hasFinalTrail$ = typeof hasFinalTrail === 'function' ? hasFinalTrail() : hasFinalTrail;
  const renderLastItemWithoutLink$ = typeof renderLastItemWithoutLink === 'function'
    ? renderLastItemWithoutLink()
    : renderLastItemWithoutLink;
  const {
    StartingTrail,
    Separator,
    Wrapper,
    Item,
    Title,
    FinalTrail,
  } = components;
  const items$ = items.map((item: CleanBreadcrumbItemType, index: number) => {
    const isLastItem = index === (items.length - 1);
    const { position, isCurrentPage } = item;
    // increment position by 1 if there is starting trail item
    const position$ = hasStartingTrail$ ? position + 1 : position;
    if (isLastItem && renderLastItemWithoutLink$) {
      const TitleWithNoLink = withDesign({
        Link: replaceWith(Fragment),
      })(Title);
      return (
        <React.Fragment key={item.uuid}>
          <Item position={position$} isCurrentPage={isCurrentPage}>
            <ItemNodeProvider nodeKey={item.nodeKey} nodeCollection={item.nodeCollection}>
              <TitleWithNoLink />
            </ItemNodeProvider>
          </Item>
        </React.Fragment>
      );
    }
    return (
      <React.Fragment key={item.uuid}>
        <Item position={position$} isCurrentPage={isCurrentPage}>
          <ItemNodeProvider nodeKey={item.nodeKey} nodeCollection={item.nodeCollection}>
            <Title />
          </ItemNodeProvider>
        </Item>
        {!isLastItem && <Separator key={`separator${item.uuid}`} />}
      </React.Fragment>
    );
  });
  const finalTrailItemPosition = (hasStartingTrail$ ? 1 : 0) + items$.length + 1;
  return (
    <Wrapper>
      { hasStartingTrail$
        && (
        <>
          <Item position={1} isCurrentPage={false} key="startingTrail">
            <StartingTrail />
          </Item>
          { (items$.length > 0 || hasFinalTrail$)
            && <Separator key="startingTrailSeparator" />}
        </>
        )}
      {items$}
      { hasFinalTrail$
        && (
        <>
          { items$.length > 0
            && <Separator key="finalTrailSeparator" />}
          <Item key="finalTrail" position={finalTrailItemPosition} isCurrentPage={false}>
            <FinalTrail />
          </Item>
        </>
        )}
    </Wrapper>
  );
};

const BreadcrumbStartComponents: BreadcrumbsComponents = {
  StartingTrail: MenuTitle,
  Separator: asComponent('span'),
  Wrapper: asComponent('ul'),
  Item: flowRight(
    withoutProps(['position', 'isCurrentPage']),
    asComponent,
  )('li'),
  Title: MenuTitle,
  FinalTrail: MenuTitle,
};

/**
 * Clean component that renders breadcrumbs.
 * @see BreadcrumbsComponents for a list of design components.
 */
const BreadcrumbsClean = designable(BreadcrumbStartComponents, 'Breadcrumbs')(BreadcrumbsClean$);

/**
 * HOC that enables rendering of starting trail for a breadcrumb based component.
 * @param Component a breadcrumb based component
 */
const withStartingTrail = addProps({
  hasStartingTrail: true,
});

/**
 * HOC that disables rendering of starting trail for a breadcrumb based component.
 * @param Component a breadcrumb based component
 */
const withoutStartingTrail = addProps({
  hasStartingTrail: false,
});

/**
 * HOC that enables rendering of final trail for a breadcrumb based component.
 * @param Component a breadcrumb based component
 */
const withFinalTrail = addProps({
  hasFinalTrail: true,
});

/**
 * HOC that disables rendering of final trail for a breadcrumb based component.
 * @param Component a breadcrumb based component
 */
const withoutFinalTrail = addProps({
  hasFinalTrail: false,
});

/**
 * HOC that adds breadcrumb props retrieved from breadcrumb store.
 */
const asBreadcrumbs = flowRight(
  observer,
  withBreadcrumbItemsFromStore,
  asStylableBreadcrumbs,
  withBreadcrumbsSD,
);

/**
 * Component that renders breadcrumb items retrieved from breadcrumb store.
 */
const MenuBreadcrumbs = asBreadcrumbs(BreadcrumbsClean);

export {
  asBreadcrumbs,
  BreadcrumbsClean,
  MenuBreadcrumbs,
  withStartingTrail as withBreadcrumbStartingTrail,
  withoutStartingTrail as withoutBreadcrumbStartingTrail,
  withFinalTrail as withBreadcrumbFinalTrail,
  withoutFinalTrail as withoutBreadcrumbFinalTrail,
};
