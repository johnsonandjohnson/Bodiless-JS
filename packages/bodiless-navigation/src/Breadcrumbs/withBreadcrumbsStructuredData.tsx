/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React, { ComponentType } from 'react';
import { Helmet } from 'react-helmet';

import { firstItemHomeLinkReducer } from './withBreadcrumbItemsFromStore';
import { useBreadcrumbStore } from './BreadcrumbStoreProvider';

import type { BreadcrumbsProps } from './types';

// We need a full url in the breadcrumbs LD item.
const generateUrl = (url: string) => (
  typeof window !== 'undefined'
    ? new URL(url, window.location.origin).href
    : url
);

type LDItemType = {
  '@type': string,
  position: number,
  item: {
    '@id': string,
    name: string | object,
  },
};

const withBreadcrumbsSD = (Component: ComponentType<BreadcrumbsProps>) => (
  props: BreadcrumbsProps,
) => {
  const {
    itemsReducer = firstItemHomeLinkReducer,
    renderLastItemWithoutLink = false,
    hasStartingTrail = false,
  } = props;
  const store = useBreadcrumbStore();
  const hasStartingTrail$ = typeof hasStartingTrail === 'function' ? hasStartingTrail() : hasStartingTrail;
  const renderLastItemWithoutLink$ = typeof renderLastItemWithoutLink === 'function'
    ? renderLastItemWithoutLink()
    : renderLastItemWithoutLink;

  // Do nothing if there is no store
  if (!store) return <Component {...props} />;

  const items = itemsReducer(store.breadcrumbTrail, { hasStartingTrail: hasStartingTrail$ });

  // Do nothing if there are no items in the store
  if (items.length <= 0) return <Component {...props} />;

  /* eslint-disable @typescript-eslint/indent */
  const breadcrumbItems = items.map(uuid => store.getItem(uuid))
    .reduce<LDItemType[]>((prev, current, index) => {
      if (current === undefined) return prev;
      const isLastItem = index === (items.length - 1);
      if (isLastItem && renderLastItemWithoutLink$) return prev;
      prev.push({
        '@type': 'ListItem',
        // We increment in 1 to accomodate for the index offset ( starts from 0 )
        position: index + 1,
        item: {
          '@id': current.link.data ? generateUrl(current.link.data) : '',
          name: current.title.data,
        },
      });
      return prev;
    },
  []);

  // We exclude first home trail and the last trail if `renderLastItemWithoutLink` enabled.
  // If no breadcrumbs left, do nothing.
  if (breadcrumbItems.length <= 0) return <Component {...props} />;

  const breadcrumbsSDHeader = {
    '@context': 'https://schema.org',
    '@type': 'BreadcrumbList',
    itemListElement: [...breadcrumbItems],
  };

  return (
    <>
      <Helmet>
        <script type="application/ld+json">
          {JSON.stringify(breadcrumbsSDHeader)}
        </script>
      </Helmet>
      <Component {...props} />
    </>
  );
};

export default withBreadcrumbsSD;
