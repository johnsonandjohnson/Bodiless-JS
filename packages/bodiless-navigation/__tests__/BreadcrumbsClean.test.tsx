/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ComponentType } from 'react';
import { mount } from 'enzyme';
import { asReadOnly } from '@bodiless/core';
import { withDesign, addProps, withoutProps } from '@bodiless/fclasses';
import { flowRight } from 'lodash';
import { BreadcrumbsClean, withBreadcrumbStartingTrail, withoutBreadcrumbFinalTrail } from '../src/Breadcrumbs';
import { withMenuTitleEditors } from '../src';

const withAttrRename = (oldAttr: string, newAttr: string) => (Component: ComponentType) => {
  const WithAttrRename = (props: any) => {
    const { [oldAttr]: oldAttrVal, ...rest } = props;
    const props$ = {
      ...rest,
      [newAttr]: oldAttrVal,
    };
    return <Component {...props$} />;
  };
  return WithAttrRename;
};

const TestBreadcrumbs = flowRight(
  withMenuTitleEditors(undefined, asReadOnly),
)(BreadcrumbsClean);

describe('BreadcrumbsClean', () => {
  it('renders as empty unordered list by default', () => {
    const wrapper = mount(<TestBreadcrumbs />);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('allows adding starting trail and design it using design api', () => {
    const Breadcrumb = flowRight(
      withDesign({
        StartingTrail: addProps({
          'aria-label': 'starting-trail',
        }),
      }),
      withBreadcrumbStartingTrail,
    )(TestBreadcrumbs);
    const wrapper = mount(<Breadcrumb />);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it.skip('allows passing items as a prop', () => {
    const items = [
      {
        uuid: Math.random(),
        link: {
          nodeKey: 'linkNodeKey1',
        },
        title: {
          nodeKey: 'titleNodeKey1',
        },
      },
      {
        uuid: Math.random(),
        link: {
          nodeKey: 'linkNodeKey1',
        },
        title: {
          nodeKey: 'titleNodeKey1',
        },
      },
    ];
    const Breadcrumb = flowRight(
      withDesign({
        Link: flowRight(
          withAttrRename('nodeKey', 'href'),
          withoutProps('nodeCollection'),
        ),
        Title: flowRight(
          withAttrRename('nodeKey', 'aria-label'),
          withoutProps('nodeCollection'),
        ),
      }),
    )(TestBreadcrumbs);
    const wrapper = mount(<Breadcrumb items={items} />);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it.skip('allows designing separator using design api', () => {
    const items = [
      {
        uuid: Math.random(),
        link: {
          nodeKey: 'linkNodeKey1',
        },
        title: {
          nodeKey: 'titleNodeKey1',
        },
      },
      {
        uuid: Math.random(),
        link: {
          nodeKey: 'linkNodeKey1',
        },
        title: {
          nodeKey: 'titleNodeKey1',
        },
      },
    ];
    const Breadcrumb = flowRight(
      withDesign({
        Link: flowRight(
          withAttrRename('nodeKey', 'href'),
          withoutProps('nodeCollection'),
        ),
        Title: flowRight(
          withAttrRename('nodeKey', 'aria-label'),
          withoutProps('nodeCollection'),
        ),
        Separator: addProps({
          className: 'separator',
        }),
      }),
    )(TestBreadcrumbs);
    const wrapper = mount(<Breadcrumb items={items} />);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it.skip('does not render separator after last item when final trail is not set', () => {
    const items = [
      {
        uuid: Math.random(),
        link: {
          nodeKey: 'linkNodeKey1',
        },
        title: {
          nodeKey: 'titleNodeKey1',
        },
      },
    ];
    const Breadcrumb = flowRight(
      withDesign({
        Link: flowRight(
          withAttrRename('nodeKey', 'href'),
          withoutProps('nodeCollection'),
        ),
        Title: flowRight(
          withAttrRename('nodeKey', 'aria-label'),
          withoutProps('nodeCollection'),
        ),
      }),
      withoutBreadcrumbFinalTrail,
    )(TestBreadcrumbs);
    const wrapper = mount(<Breadcrumb items={items} />);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
