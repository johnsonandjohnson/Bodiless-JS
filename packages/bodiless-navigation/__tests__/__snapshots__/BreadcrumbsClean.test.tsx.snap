// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadcrumbsClean allows adding starting trail and design it using design api 1`] = `"<nav><ol><li><a aria-label=\\"starting-trail\\"></a></li></ol></nav>"`;

exports[`BreadcrumbsClean allows designing separator using design api 1`] = `"<nav><ol><li><a href=\\"linkNodeKey1\\"><span aria-label=\\"titleNodeKey1\\"></span></a></li><span class=\\"separator\\"></span><li><a href=\\"linkNodeKey1\\"><span aria-label=\\"titleNodeKey1\\"></span></a></li></ol></nav>"`;

exports[`BreadcrumbsClean allows passing items as a prop 1`] = `"<nav><ol><li><a href=\\"linkNodeKey1\\"><span aria-label=\\"titleNodeKey1\\"></span></a></li><span></span><li><a href=\\"linkNodeKey1\\"><span aria-label=\\"titleNodeKey1\\"></span></a></li></ol></nav>"`;

exports[`BreadcrumbsClean does not render separator after last item when final trail is not set 1`] = `"<nav><ol><li><a href=\\"linkNodeKey1\\"><span aria-label=\\"titleNodeKey1\\"></span></a></li></ol></nav>"`;

exports[`BreadcrumbsClean renders as empty unordered list by default 1`] = `"<nav><ol></ol></nav>"`;
